Move parse tree binding from optimizer to catalog

nappelson opened the pull request:
Goal was to refactor the optimizer to pull out the parse tree binding logic as it doesn't belong in the optimizer code. 

In future changes, I'd like to add some utilities to make it easier to modify our tests so that we don't need to touch all of the tests below when making a change like this.

chenboy commented on test/executor/copy_test.cpp:
```diff
@@ -120,8 +120,7 @@ TEST_F(CopyTests, Copying) {
   LOG_TRACE("Building plan tree...");
-  auto copy_plan =
-      optimizer->BuildPelotonPlanTree(copy_stmt, DEFAULT_DB_NAME, txn);
+  auto copy_plan = optimizer->BuildPelotonPlanTree(copy_stmt, txn);
```
Do we want to bind the parse tree first in this test?

nappelson replied:
Yes done thanks. I am actually curious why this test passes without binding.

apavlo commented:
@nappelson Can you rebase correctly? It looks like you have duplicate commits.

nappelson commented:
Yep just fixed the issue thanks - sorry about that.

On Sun, Feb 11, 2018 at 9:47 PM, Andy Pavlo <notifications@github.com>
wrote:

> @nappelson <https://github.com/nappelson> Can you rebase correctly? It
> looks like you have duplicate commits.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cmu-db/peloton/pull/1133#issuecomment-364817445>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADoPvm6yeTtWELcjQj0qOL9G0ln93r5tks5tT6ZMgaJpZM4R9waQ>
> .
>

pmenon commented on src/catalog/abstract_catalog.cpp:
```diff
@@ -10,11 +10,12 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include "binder/bind_node_visitor.h"
```
Follow the style guide for the order of includes [here](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes) (i.e., the first include should be `catalog/abstract_catalog.h` in `catalog/abstract_catalog.cpp`).

pmenon commented on src/catalog/abstract_catalog.cpp:
```diff
@@ -37,6 +38,8 @@
 #include "storage/storage_manager.h"
 #include "storage/table_factory.h"
 
+using std::make_shared;
```
Don't use these `using` directives.

nappelson replied:
@pmenon thanks! just pushed these changes. I have avoided the use of `using` directives in other projects but noticed that the codebase has a fair amount of these. I will do my best to refactor them out in the future for other diffs.

nappelson commented:
@apavlo @pmenon sorry about that - now there is no more `using`.

nappelson commented:
@pmenon just pinging again as this should be ready now.

pmenon commented:
It looks good overall. General questions:
1. It seems like `BindNodeVisitor` is a single-use class that doesn't outlive its scope of use. If so, does it make sense to allocate on stack rather than heap-allocate with `std::make_shared`?
2. `BindNodeVisitor` exposes an implementation detail (i.e., visitor pattern). Does it make sense to have a generic `Binder` class that implements binding with `BindNodeVisitor`?

mengranwo commented:
@nappelson 
Hi, can you take a look at this PR and @pmenon's comments? If you have addressed his requests, please change the tag into "accept" 

Thanks!

nappelson commented:
@pmenon just pushed change addressing your comment regarding stack allocation for BindNodeVisitor. 

For your second comment, that makes sense however the BindNodeVisitor currently implements SqlNodeVisitor (which also exposes an implementation detail). I think a larger refactor like this should be another explicit task to help keep the PRs smaller. 

What are your thoughts?

nappelson commented:
@pmenon just wanted to check in again

apavlo commented:
@nappelson We need to come back around and get this in. Can you resolve the conflicts and make sure you address all of the comments?

@malin1993ml Can you work with Nate to get this finished? I think Prashanth is tied up with other things. Thanks

nappelson commented:
Yep just resolved the conflicts. I responded to one of Prashanth's comments
a while back and checked in but haven't heard back from him. It says the
merge is blocked until I address the requested changes (I am assuming
either Prashanth has to review it again or he needs to be removed as a
reviewer).

Best,
Nate

On Mon, Apr 16, 2018 at 12:44 PM, Andy Pavlo <notifications@github.com>
wrote:

> @nappelson <https://github.com/nappelson> We need to come back around and
> get this in. Can you resolve the conflicts and make sure you address all of
> the comments?
>
> @malin1993ml <https://github.com/malin1993ml> Can you work with Nate to
> get this finished? I think Prashanth is tied up with other things. Thanks
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cmu-db/peloton/pull/1133#issuecomment-381669975>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADoPvu5sMp617yPgG2kCjlUUjEU0nlpsks5tpMpmgaJpZM4R9waQ>
> .
>

malin1993ml commented:
Hi @crd477 @pervazea, it looks like some of our PRs including this one cannot download the repo with the Jenkins build again. For example:
http://jenkins.db.cs.cmu.edu:8080/blue/organizations/jenkins/peloton/detail/PR-1322/1/pipeline/
http://jenkins.db.cs.cmu.edu:8080/blue/organizations/jenkins/peloton/detail/PR-1321/1/pipeline/
http://jenkins.db.cs.cmu.edu:8080/blue/organizations/jenkins/peloton/detail/PR-1133/86/pipeline/
http://jenkins.db.cs.cmu.edu:8080/blue/organizations/jenkins/peloton/detail/PR-1310/7/pipeline/

I'm wondering is this because we have another machine that has networking problem? Would really appreciate your help.

crd477 commented:
Ugh...

Yes, this was the default route problem again that originally cropped up in #1295.  I thought I'd fixed it the other day but obviously it happened again.  I've implemented a more permanent fix now and it should not happen again.

Sorry (again) for the inconvenience...
